src/rsh/mod.rs

```diff
@@ -1,30 +1,62 @@
 pub mod builtins;
+pub mod utils;
 
-use std::collections::HashMap;
+use std::env;
 use std::path::PathBuf;
+
 use std::io;
-use std::io::{Read, Write};
+use std::io::Write;
+
+use std::collections::HashMap;
 use std::collections::hash_map::Entry;
 
-#[derive(PartialEq, Debug, Clone)]
+#[derive(Debug, Clone)]
 pub struct State {
     cwd: PathBuf,
-    environment: HashMap<String, String>,
+    variables: HashMap<String, String>,
     aliases: HashMap<String, String>,
     argv: Vec<String>,
     argc: usize,
+    exit_status: i32,
 }
 
 impl State {
     pub fn new(cwd: String) -> State {
         State {
-            cwd: PathBuf::from(cwd),
-            environment: HashMap::new(),
+            cwd: utils::make_absolute(PathBuf::from(cwd)).unwrap(),
+            variables: HashMap::new(),
             aliases: HashMap::new(),
             argv: Vec::new(),
             argc: 0,
+            exit_status: 0,
+        }
+    }
+
+    pub fn default() -> State {
+        match env::current_dir() {
+            Ok(cwd) => {
+                State {
+                    cwd: cwd,
+                    variables: HashMap::new(),
+                    aliases: HashMap::new(),
+                    argv: Vec::new(),
+                    argc: 0,
+                    exit_status: 0,
+                }
+            }
+            Err(e) => panic!(e),
         }
     }
+
+    pub fn env<'a>(&'a mut self, key: String, value: String) -> &'a mut State {
```

Comment:
This is using [lifetimes](https://doc.rust-lang.org/book/lifetimes.html) which are part of the ownership  / borrowing system in Rust. What is says is that whatever the lifetime of self is is what the lifetime of the reference we are passing back should be. It tells the compiler that whatever code block self is part of (i.e. when it will be dropped / deallocated) is the same time that the reference we pass back should be dropped / deallocated.

Lifetimes are super hard and I would recommend reading the linked docs above and these blog posts:

http://www.charlesetc.com/rust/2015/10/29
https://medium.com/@ericdreichert/how-i-think-about-rust-lifetimes-83a726aaa846#.x35zzejno